llvm/lib/Transforms/Scalar/ConstraintElimination.cpp

```diff
@@ -269,6 +269,19 @@ static bool eliminateConstraints(Function &F, DominatorTree &DT) {
       continue;
     WorkList.emplace_back(DT.getNode(&BB));
 
+    // Returns true if we can add a known condition from BB to its successor
+    // block Succ. Each predecessor of Succ can either be BB or be dominated by
+    // Succ (e.g. the case when adding a condition from a pre-header to a loop
+    // header).
+    auto CanAdd = [&BB, &DT](BasicBlock *Succ) {
+      assert(isa<BranchInst>(BB.getTerminator()));
+      return any_of(successors(&BB),
+                    [Succ](const BasicBlock *S) { return S != Succ; }) &&
```

Comment:
With this `any_of...`, `CanAdd` returns `false`, if `Succ` is the only successor of `BB`? For example, in the case like below, does it add the assumption for `%cond` to `%succ`?

```
%bb:
  %cond = icmp ugt i16 %a, 0
  assume(%cond)
  br %succ
%succ:
  %cond2 = icmp ugt i16 %a, 0
  return %cond2
```

Reply:
> With this any_of..., CanAdd returns false, if Succ is the only successor of BB? For example, in the case like below, does it add the assumption for %cond to %succ?

That's a good point! We would not add conditions from assumes if there is an unconditional branch as in your example. This can be easily supported and I added another commit to fix it: 12e1762 & 1821ec8